File: 1884.egg-drop-with-2-eggs-and-n-floors.kt
// @leet start
class Solution {
    fun twoEggDrop(n: Int, k: Int, dp: Array<IntArray>): Int {
      if (n == 0 || k == 0) {
        return 0
      }
      if (k == 1) {
        return n
      }
      if (dp[k][n] != 0) {
        return dp[k][n]
      }

      dp[k][n] = n
      for (i in 1..n) {
        dp[k][n] = Math.min(dp[k][n], 1 + max(twoEggDrop(i - 1, k - 1, dp), twoEggDrop(n - i, k, dp)))
      }

      return dp[k][n]
    }
    fun twoEggDrop(n: Int): Int {
      // k eggs
      var k = 2
      // two dimentional dp
      var dp = Array(k + 1) { IntArray(n + 1) }
      return twoEggDrop(n, k, dp)
    }
}
// @leet end

File: 1884.egg-drop-with-2-eggs-and-n-floors.cs
// @leet start
public class Solution {
    int[] dp;
    int Solve(int n, int k) {
      if (n == 0 || k == 0) {
        return 0;
      }
      if (k == 1) {
        return n;
      }
      if (dp[n] != 0) {
        return dp[n];
      }

      var min = n;
      for (int i = 1; i <= n; i++) {
        min = Math.Min(min, 1 + Math.Max(Solve(i - 1, k - 1), Solve(n - i, k)));
      }
      dp[n] = min;

      return dp[n];
    }
    public int TwoEggDrop(int n) {
      dp  = new int[n+1];
      return Solve(n, 2);
    }
}
// @leet end
//
// x,k
//    |   | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 |   |   |   |   |
//    | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |   |   |   |   |
//    | 1 | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 |   |   |   |   |
//    | 2 | 0 | 1 | 2 | 2 | 3 | 3 | 3 | 4 |   |   |   |   |
//    | 3 | 0 |   |   |   |   |   |   |   |   |   |   |   |
//
//
